core/src/main/java/feign/Contract.java

```diff
@@ -211,99 +207,224 @@ protected void nameParam(MethodMetadata data, String name, int i) {
+            if (type.isInstance(annotation))
+              processor.process(annotation, data);
+          }));
+    }
+
+    /**
+     * @param data metadata collected so far relating to the current java method.
+     * @param annotation annotations present on the current method annotation.
+     * @param method method currently being processed.
+     */
     @Override
-    protected void processAnnotationOnClass(MethodMetadata data, Class<?> targetType) {
-      if (targetType.isAnnotationPresent(Headers.class)) {
-        String[] headersOnType = targetType.getAnnotation(Headers.class).value();
+    protected final void processAnnotationOnMethod(MethodMetadata data,
+                                                   Annotation annotation,
+                                                   Method method) {
+      methodAnnotationProcessors.forEach((type, processor) -> {
+        if (type.isInstance(annotation))
+          processor.process(annotation, data);
+      });
+    }
+
+
+    /**
+     * @param data metadata collected so far relating to the current java method.
+     * @param annotations annotations present on the current parameter annotation.
+     * @param paramIndex if you find a name in {@code annotations}, call
+     *        {@link #nameParam(MethodMetadata, String, int)} with this as the last parameter.
+     * @return true if you called {@link #nameParam(MethodMetadata, String, int)} after finding an
+     *         http-relevant annotation.
+     */
+    protected final boolean processAnnotationsOnParameter(MethodMetadata data,
+                                                          Annotation[] annotations,
+                                                          int paramIndex) {
+      return Arrays.stream(annotations)
+          .flatMap(annotation -> parameterAnnotationProcessors.entrySet()
+              .stream()
+              .map(entry -> entry.getKey().isInstance(annotation)
+                  ? entry.getValue().process(annotation, data, paramIndex)
+                  : false))
+          .collect(Collectors.reducing(Boolean::logicalOr))
+          .orElse(false);
+    }
+
+    @FunctionalInterface
+    public interface ClassAnnotationProcessor<E extends Annotation> {
+      /**
+       * Called by parseAndValidateMetadata twice, first on the declaring class, then on the target
+       * type (unless they are the same).
+       *
+       * @param annotation present on the current class
+       * @param metadata metadata collected so far relating to the current java method.
+       */
+      void process(E annotation, MethodMetadata metadata);
+    }
+
+    /**
+     * Called while class annotations are being processed
+     *
+     * @param annotation to be processed
+     * @param processor function that defines the annotations modifies {@link MethodMetadata}
+     */
+    protected <E extends Annotation> void registerClassAnnotation(Class<E> annotation,
```

Comment:
`register*` methods modifies internal state, this is not good, maybe need to find a way to "register" processors during instance creation and then assign to a unmodifiable collection.

Reply:
We can use this as long as we make it safe.  However, I agree that the processors should be provided during construction to make the contract immutable

Reply:
Take a look at this: https://github.com/spring-cloud/spring-cloud-openfeign/blob/master/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/support/SpringMvcContract.java

This is really close to what I think you are going for.

Reply:
I created my own POC here: https://github.com/kdavisk6/feign/tree/declarative-contract-poc/core/src/main/java/feign/contract

I think we can consolidate this down into a single Map and introduce the concept of a context to wrap up the execution state.  Take a look and let me know what you think.

Reply:
I did apply this "declarative" concept to jaxrs contract... and for that, I would need something more complex to evaluate if an annotation should be processed or not.

I'm thinking about a Predicate... if I go for a predicate, then, a single map (or list) should be enough for all cases.

Reply:
I looked through the JAXRS Contract.  I may be missing something, but I did not see anything that requires the contract to conditionally parse an annotation outside of skipping ignored annotations.  Is that what you are looking into?

Reply:
For jaxrs, we don't check if @GET is present, if check if method are annotated with any annotation that in turn is annotated with @HttpMethod

Reply:
Predicate for annotationProcessors solved this